### PR TITLE
Add rule against using specific deprecated jQuery Methods (Blacklisted)

### PR DESCRIPTION
**Pull request by lynchbomb**

### CODE REVIEW

**Comment by lynchbomb** [comment]
I'm also totally cool with moving this effort/PR into a branch and keeping it there. I'm unsure if this type of linting rule is applicable to many folks? @stefanpenner

File: .gitignore [hunk 1]
```diff
@@ -16,3 +16,4 @@
 /libpeerconnection.log
 npm-debug.log
 testem.log
+.DS_Store
```
Comment (trentmwillis) [inline .gitignore]: Pro-tip: Add this to your global git config and never worry about it again

File: lib/index.js [hunk 2]
```diff
@@ -2,5 +2,6 @@ const requireindex = require('requireindex');
 
 module.exports.rules = requireindex(__dirname + '/rules');
 module.exports.configs = {
-  recommended: require('../config/recommended')
+  recommended: require('../config/recommended'),
```
Comment (trentmwillis) [inline lib/index.js]: Unrelated to this PR, should we be exporting the `experimental` config here as well?

File: lib/rules/no-jquery-methods.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    recommended: true
+  },
+  meta: {
+    blacklist: BLACKLIST
+  },
+  create(context) {
+    return {
+      MemberExpression(node) {
+        let propertyName = node.property.name;
+        let blackListName = BLACKLIST.includes(propertyName) ? propertyName : false;
```
Comment (trentmwillis) [inline lib/rules/no-jquery-methods.js]: Looking at this it seems like we'll wind up blacklisting any method that is in the blacklist whether it is part of a jQuery object or not. We should add tests to ensure that isn't the case.

Note: I feel like doing this properly will be difficult due to variables, e.g.:

```js
const myJQueryObj = this.$();
myJQueryObj.add(foo);
```

File: .vscode/launch.json [hunk 4]
```diff
@@ -0,0 +1,29 @@
+{
```
Comment (trentmwillis) [inline .vscode/launch.json]: I'm unsure we want to commit this to the project, though if we do, we should likely add a description stating what this is useful for debugging.

File: .gitignore (hunk 1 above)
Comment (lynchbomb) [inline .gitignore]: Nice. I normally just flag OSX to stop producing them but I just upgraded to Sierra which defaulted that setting. Fail.

File: lib/rules/no-jquery-methods.js (hunk 3 above)
Comment (lynchbomb) [inline lib/rules/no-jquery-methods.js]: I'll include a few tests to confirm, but yeah this is pretty tricky to properly ensure ^^^^

File: .vscode/launch.json (hunk 4 above)
Comment (lynchbomb) [inline .vscode/launch.json]: I'll pluck it out. Eventually we should create helpful plugin debugging for ESLint. Ideally something better than using http://astexplorer.net/

File: lib/index.js (hunk 2 above)
Comment (lynchbomb) [inline lib/index.js]: Yeah otherwise I don't know how folks are consuming the `experimental` rules

File: lib/rules/no-jquery-methods.js (hunk 3 above)
Comment (lynchbomb) [inline lib/rules/no-jquery-methods.js]: I've confirmed the following use cases are correctly flagged with the current rule:
```
const my$ = Ember.$;
my$.after();
```
```
const my$ = $;
my$.after();
```

The tricky bit that is currently being flagged as blacklisted, is when a custom method has the same name as a blacklisted method; we should allow for this and i'll need to write a few tests to confirm we are:
```
const myObj = {
  add(){}
}
myObj.add()
```

File: lib/rules/no-jquery-methods.js [hunk 5]
```diff
@@ -24,6 +24,14 @@ const BLACKLIST = [
   'wrapAll', 'wrapInner'
 ];
 
+function isJQueryCallee(node) {
```
Comment (lynchbomb) [inline lib/rules/no-jquery-methods.js]: @trentmwillis this will cover us accidentally blacklisting non-jQuery object methods. Ensuring we allow for at a min your example:

```
const myJQueryObj = this.$();
myJQueryObj.add(foo);
```

File: tests/lib/rules/no-jquery-methods.js [hunk 6]
```diff
@@ -15,16 +15,25 @@ ruleTester.run('no-jquery-methods', rule, {
...
             this.$().notBlacklistedMethod();
-          },
+          }
+        });`
+    },
+    {
+      code: `
+        export default Ember.Component({
+          init() {
+            const myJQueryObj = this.$();
```
Comment (lynchbomb) [inline tests/lib/rules/no-jquery-methods.js]: @trentmwillis tested per above.

**Comment by lynchbomb** [comment]
@rwjblue brought up a good point regarding `import` declarations and ensuring we are covering use-cases like:

```js
import jQuery from 'jquery';

jQuery.add()
```

I'm going to append this commit to include the updated rule and test(s) for above.

**Comment by lynchbomb** [comment]
@rwjblue brought up another good point (DM) to leverage a schema to allow folks to modify the `BLACKLIST` via `.eslintrc.js` as a param. Updated PR coming.

File: config/jquery.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
+module.exports = {
```
Comment (scalvert) [inline config/jquery.js]: Should this config be named `no-jquery`?

File: lib/rules/no-jquery-methods.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+ */
+'use strict';
+
+/**
+ * Determines if this expression matches the method blacklist invocation.
+ * @param {ASTNode} node The identifier node.
+ * @returns {Boolean} Returns true if the expression matches, otherwise false.
+ */
+
+const BLACKLIST = [
```
Comment (scalvert) [inline lib/rules/no-jquery-methods.js]: This list would be easier to read if each were on a separate line. Can we do this?

File: lib/rules/no-jquery-methods.js [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  'wrapAll', 'wrapInner'
+];
+
+function isJQueryCaller(node, name = null) {
+  let oName = node.object && node.object.callee && node.object.callee.object && node.object.callee.object.name === 'Ember';
+  let pName = node.object && node.object.callee && node.object.callee.property && node.object.callee.property.name === '$';
+  let cName = node.object && node.object.callee && node.object.callee.name === '$';
+  let lName = node.object && node.object.name === name;
+
+  return (oName) || (pName) || (cName) || (lName) ? true : false;
```
Comment (scalvert) [inline lib/rules/no-jquery-methods.js]: Can we just trim off the ternary and just return the result of the ORs?

Comment (scalvert) [inline lib/rules/no-jquery-methods.js]: Also, parens unnecessary, no?

File: lib/rules/no-jquery-methods.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+  meta: {
+    blacklist: BLACKLIST
+  },
+  create(context) {
+    let isJQuerySource = null;
+    let jQueryLocalName = null;
+
+    return {
+      ImportDeclaration(node) {
+        return isJQuerySource = node.source && node.source.value === 'jquery';
```
Comment (scalvert) [inline lib/rules/no-jquery-methods.js]: Y u no use the `import` util?

File: lib/rules/no-jquery-methods.js [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+      VariableDeclarator(node) {
+        return jQueryLocalName = jQueryLocalName === node.init.name ? node.id.name : null;
+      },
+      MemberExpression(node) {
+        let propertyName = node.property.name;
+        let blackListName = BLACKLIST.includes(propertyName) ? propertyName : false;
+        let isThisExpression = node.object && node.object.type === 'ThisExpression';
+
+        if (!isThisExpression && blackListName && isJQueryCaller(node, jQueryLocalName)) {
+          context.report({ node: node.property, message: `The use of ${blackListName} method has been deprecated.` });
```
Comment (scalvert) [inline lib/rules/no-jquery-methods.js]: Can we extract the message into a `const`, similar to how we do in the other rules? We do this so we can reuse in our tests. You'll also need to add it to `meta` as a property, i.e.

```js
meta: {
  blacklist: BLACKLIST,
  message: MESSAGE
}
```

File: tests/lib/rules/no-jquery-methods.js [hunk 12]
```diff
@@ -0,0 +1,100 @@
+const rule = require('../../../lib/rules/no-jquery-methods');
+const MESSAGE = rule.meta.message;
```
Comment (scalvert) [inline tests/lib/rules/no-jquery-methods.js]: This is currently `undefined`. You'll need to add message to your rule's meta.

File: tests/lib/rules/no-jquery-methods.js [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+            this.$().notBlacklistedMethod();
+          }
+        });`
+    },
+    {
+      code: `
+        export default Ember.Component({
+          init() {
+            const myJQueryObj = this.$();
+            myJQueryObj[${BLACKLISTMETHOD}]();
```
Comment (scalvert) [inline tests/lib/rules/no-jquery-methods.js]: Shouldn't this be invalid? We want the rule to fire if the method is blacklisted, no?

File: config/jquery.js (hunk 7 above)
Comment (lynchbomb) [inline config/jquery.js]: Cool with me. I'll rename.

File: lib/rules/no-jquery-methods.js (hunk 9 above)
Comment (lynchbomb) [inline lib/rules/no-jquery-methods.js]: Both are totally for readability (old-habit) and are options/can be removed.

File: lib/rules/no-jquery-methods.js (hunk 10 above)
Comment (lynchbomb) [inline lib/rules/no-jquery-methods.js]: Overkill. All this cares about is the value of one key (`node.source.value`).

File: lib/rules/no-jquery-methods.js (hunk 11 above)
Comment (lynchbomb) [inline lib/rules/no-jquery-methods.js]: It appears when a message is passed into `context.report` it is included within `meta.message` thus doing above. To confirm in the correlating test I am grabbing `rule.meta.message`.

I intentionally didn't do this because the message we are reporting is contextual and not static (blackListName is included in the message).

Comment (lynchbomb) [inline lib/rules/no-jquery-methods.js]: I had this originally, however the message we are reporting is contextual not static (including the `blackListName` within the message). I'll look into converting this to a function if possible.

File: tests/lib/rules/no-jquery-methods.js (hunk 12 above)
Comment (lynchbomb) [inline tests/lib/rules/no-jquery-methods.js]: Nice catch!

File: tests/lib/rules/no-jquery-methods.js (hunk 13 above)
Comment (lynchbomb) [inline tests/lib/rules/no-jquery-methods.js]: There is a previous discussion about this between @trentmwillis and I. We are intentionally flagging this as `valid`; however I am second guessing this. Essentially why would we ever want to allow a jQuery blacklisted method thru?

I am in agreement (with you) that we should modify this to be invalid. @trentmwillis additional thoughts?

File: lib/rules/no-jquery-methods.js (hunk 10 above)
Comment (scalvert) [inline lib/rules/no-jquery-methods.js]: But don't we care that we can alias the import to something other than `jQuery`?

File: lib/rules/no-jquery-methods.js (hunk 11 above)
Comment (scalvert) [inline lib/rules/no-jquery-methods.js]: Makes sense. FWIW, I do this in https://github.com/scalvert/ember-bpr-lint/blob/master/lib/rules/no-unguarded-globals.js#L44

Comment (lynchbomb) [inline lib/rules/no-jquery-methods.js]: Perfect. TY for the example.

File: lib/rules/no-jquery-methods.js (hunk 10 above)
Comment (lynchbomb) [inline lib/rules/no-jquery-methods.js]: We are checking the alias against the import within the `ImportDefaultSpecifier`. On the chance folks have a different alias ie `$, jQ` etc.

```js
ImportDefaultSpecifier(node) {
  return jQueryLocalName = isJQuerySource ? node.local.name : null;
}
```

File: lib/rules/no-jquery-methods.js (hunk 8 above)
Comment (lynchbomb) [inline lib/rules/no-jquery-methods.js]: I'm going to remove the hardcoded `const BLACKLIST...` from the rule and instead have the BLACKLIST passed as a param `.eslintrc.js` (context.options)

File: lib/index.js (hunk 2 above)
Comment (scalvert) [inline lib/index.js]: Are you going to add this to this PR?

File: lib/rules/no-jquery-methods.js (hunk 10 above)
Comment (scalvert) [inline lib/rules/no-jquery-methods.js]: Makes sense.

File: tests/lib/rules/no-jquery-methods.js (hunk 13 above)
Comment (scalvert) [inline tests/lib/rules/no-jquery-methods.js]: Are we going to make this change in this PR?

Comment (trentmwillis) [inline tests/lib/rules/no-jquery-methods.js]: My original point was that this should be invalid, but detecting it would be very difficult as you'll have to walk back the reference or keep track of all variables bound to jquery values

Comment (trentmwillis) [inline tests/lib/rules/no-jquery-methods.js]: To add a bit more color to that last comment. I assumed this test was being marked as valid to acknowledge that we aren't specifically trying to catch usage from variables. I say this because doing it well will be very difficult. Consider:

```js
function show() {
  const el = getElement();
  el.show();
}
```

Or even:

```js
Component.extend({
  init() {
    this.input = this.$('input');
  },
  toggleInput() {
    this.input.toggle();
  }
});
```

If we do try to catch any of these, we should be explicit about what we can and cannot reasonably detect so that consumers understand the behavior accurately.

File: lib/index.js (hunk 2 above)
Comment (lynchbomb) [inline lib/index.js]: I was going to submit a separate PR for this and batch it with a change in the `no-global-jquery` rule @rwjblue noticed (in the rule we are only checking for global `$` and not also `jQuery`.

File: tests/lib/rules/no-jquery-methods.js (hunk 13 above)
Comment (lynchbomb) [inline tests/lib/rules/no-jquery-methods.js]: Thanks for the detailed response @trentmwillis. Total misunderstanding on my part. I'll update that test to be invalid and additionally i'll include a test for your example above.

Comment (scalvert) [inline tests/lib/rules/no-jquery-methods.js]: Ya I totally agree this is an issue. I think this is a massive slippery slope, and I for the most part what we have is addressing the bulk of what we need. 

We should merge this guy and address additional things as they crop up.

File: config/no-jquery.js [hunk 14]
```diff
@@ -0,0 +1,31 @@
+const BLACKLIST = [
```
Comment (scalvert) [inline config/no-jquery.js]: Can we add newlines for each of these? It makes it more readable.

File: guides/rules/no-jquery-methods.md [hunk 15]
```diff
@@ -0,0 +1,6 @@
+# No jQuery Methods
+
+**TL;DR** The use of the specific blacklist jQuery method has been deprecated.
```
Comment (scalvert) [inline guides/rules/no-jquery-methods.md]: "The intent of this rule is to identify any blacklisted jQuery methods as deprecated"

File: lib/rules/no-jquery-methods.js [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+ * @param {ASTNode} node The identifier node.
+ * @returns {Boolean} Returns true if the expression matches, otherwise false.
+ */
+
+function getMessage(blackListName) {
+  return `The use of ${blackListName} method has been deprecated.`;
+}
+
+function isJQueryCaller(node, name = null) {
+  let oName = node.object && node.object.callee && node.object.callee.object && node.object.callee.object.name === 'Ember';
```
Comment (scalvert) [inline lib/rules/no-jquery-methods.js]: I think you can use the `get` util here.

File: .gitignore (hunk 1 above)
Comment (scalvert) [inline .gitignore]: Let's just get rid of this :)

File: guides/rules/no-jquery-methods.md (hunk 15 above)
Comment (lynchbomb) [inline guides/rules/no-jquery-methods.md]: C

File: config/no-jquery.js (hunk 14 above)
Comment (lynchbomb) [inline config/no-jquery.js]: C

File: .gitignore (hunk 1 above)
Comment (lynchbomb) [inline .gitignore]: C

File: lib/rules/no-jquery-methods.js [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+ */
+'use strict';
+
+/**
+ * Determines if this expression matches the method blacklist invocation.
+ * @param {ASTNode} node The identifier node.
+ * @returns {Boolean} Returns true if the expression matches, otherwise false.
+ */
+
+function get(node, path) {
```
Comment (scalvert) [inline lib/rules/no-jquery-methods.js]: Is this duplicate code from our utils that was copypasta-ed into this when transporting form astexplorer? If so, can we pew pew?

🔫

File: lib/rules/no-jquery-methods.js [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+    let varDecIDName = null;
+
+    return {
+      ImportDeclaration(node) {
+        let localName = getLocalImportName(node, 'jquery');
+        if (localName.length > 0) {
+          jQueryLocalName = localName[0];
+        }
+      },
+      VariableDeclarator(node) {
```
Comment (scalvert) [inline lib/rules/no-jquery-methods.js]: This is a bit of a nit, but it goes toward readability, but can we add a blank line in between these members?

```js
return {
  ImportDeclaration(node) {
    let localName = getLocalImportName(node, 'jquery');
    if (localName.length > 0) {
       jQueryLocalName = localName[0];
    }
  },

VariableDeclarator(node) {
...
```

File: lib/rules/no-jquery-methods.js (hunk 17 above)
Comment (lynchbomb) [inline lib/rules/no-jquery-methods.js]: Not from astexplorer. I'm following this similar pattern from other rules within this repo. 

I think it provides some value.

File: lib/rules/no-jquery-methods.js (hunk 18 above)
Comment (lynchbomb) [inline lib/rules/no-jquery-methods.js]: C

File: lib/rules/no-jquery-methods.js (hunk 17 above)
Comment (scalvert) [inline lib/rules/no-jquery-methods.js]: Oh, but don't we want to use `lib/utils/get.js`?

File: lib/rules/no-jquery-methods.js [hunk 19]
```diff
@@ -0,0 +1,76 @@
+/**
+ * @fileOverview Disallow the use of specific jQuery methods.
+ * @author Marc Lynch
+ */
+'use strict';
+
+/**
```
Comment (scalvert) [inline lib/rules/no-jquery-methods.js]: Is this comment in the wrong place?

**Comment by scalvert** [comment]
LGTM 👍
